README.md
# Rest for a while

> New download address [https://marketplace.visualstudio.com/items?itemName=jeffwong.rest-for-a-while](https://marketplace.visualstudio.com/items?itemName=jeffwong.rest-for-a-while)

Rest is a extension that reminding you do not work too hard. Be careful about your health.

## Usage

+ Press  `Ctrl + Shift+ P`
+ Enter `install`
+ Search `RestWhile`
+ Install it

It will remind you have a break when you continue work for 60 minutes. You can also set the time as you want.

When you leave for 10 minutes, your continuous working time will be zeroed.

You can also see your accumulated time, it's the sum of several continuous working time

## Command
+ Press `Ctrl + Shift+ P`
+ Enter `Rest:showContinuousTime`
You can see how long have you continuous worked!
+ Enter `Rest:showAccumulatedTime`
You can see how long have you


CHANGELOG.md
## 0.0.6
- fix the bug: tip remind error

## 0.0.3
- Add locale support
- Add support for 'showAccumulatedTime'
- Add settings `isShowTips`
- rebuild code structure


extension.js
let { window, workspace, Disposable, StatusBarAlignment, commands } = require('vscode');

const EXTENSIONNAME = "rest";

class RestController {

    constructor() {
        let subscriptions = [];
        this.initConfig();
        this._startTime = this._getCurrentTime(); // 编辑器开始启动的时间
        this._lastTime = this._startTime; // 当前时间
        this._nowTime = this._lastTime; // 上一次触发的时间
        this._continueTime = 0;
        this._nextTime = this._codingTime; // 预计下次提醒时间
        this._statusItem = window.createStatusBarItem(StatusBarAlignment.Left, 3);
        window.onDidChangeTextEditorSelection(this.onPopTips, this, subscriptions);
        window.onDidChangeActiveTextEditor(this.onPopTips, this, subscriptions);
        this._disposable = Disposable.from(...subscriptions);
        console.log(workspace.workspaceFolders)
        console.log(workspace.name)

    }

    changeStatusBarItem(minutes) {
        if(!this._statusItem) {
            this._statusItem = window.createStatusBarItem(StatusBarAlignment.Left, 3);
        }
        this._statusItem.text = `编程时长 ${minutes} 分钟`;
        this._statusItem.show();
    }

    popTips(minutes) {
        let hours = Math.floor(minutes/60);
        minutes = minutes % 60;
        let tmp = hours === 0 ? "" : (hours+"小时");
        window.showWarningMessage(`你已经编程超过${tmp}${minutes}分钟! 休息一会吧!`);
    }

    onPopTips(event) {
        console.log(event)
        this._lastTime = this._nowTime;
        this._nowTime = this._getCurrentTime();
        if(this._nowTime - this._lastTime >= this._deadTime ) {
            this._startTime = this._nowTime;
            this._nextTime = this._codingTime;
            this._continueTime = 0;
            return;
        }
        this._continueTime = this._nowTime - this._startTime;
        let minutes = this._reverseTime(this._continueTime, false);
        this.changeStatusBarItem(minutes);
        if(this._continueTime >= this._nextTime) {
            this._nextTime += this._codingTime;
            this.popTips(minutes);

        }
    }
    getContinueTime() {
        this._lastTime = this._nowTime;
        this._nowTime = this._getCurrentTime();
        this._continueTime = this._nowTime - this._startTime;
        return this._continueTime;
    }
    _getConfigTime(value) {
        return this._reverseTime(this._config.get(value));
    }
    // 时间转换，true是分钟转毫秒，false是毫秒转分
    _reverseTime(time, to) {
        if(to || to === undefined) {
            return time*60000;
        } else {
            return Math.floor(time/60000);
        }
    }

    initConfig() {
        this._config = workspace.getConfiguration(EXTENSIONNAME);
        if(!this._config) return;
        this._codingTime = this._getConfigTime("interval");
        if(!this._codingTime) return;
        this._deadTime = this._getConfigTime("departure");
        if(!this._deadTime) return;
    }

    _getCurrentTime() {
        return new Date().getTime();
    }

    dispose() {
        this._disposable.dispose();
    }

}


function activate(context) {
    let controller = new RestController();
    // 用户配置更改时，初始化控制器的配置
    workspace.onDidChangeConfiguration(()=>
        controller.initConfig());
    context.subscriptions.push(controller,
    commands.registerCommand('rest.showContinousTime', function(){
        let m = controller._reverseTime(controller.getContinueTime(), false);
        window.showInformationMessage(`当前编程时长为: ${m}分钟`);
    }));
}
exports.activate = activate;

function deactivate() {

}
exports.deactivate = deactivate;


src/store.js
const { getCurrentDate, getCurrentTime }  =  require('./utils')
const _ = require('lodash')

let Store = {}

exports.setAll = function (state) {
  Store = state
  if (!Store.lastTime) {
    Store.lastTime = getCurrentTime()
  }
  if (!Store.records) {
    Store.records = {}
  }
}

exports.getAll = function () {
  return Store
}

exports.updateLastTime = function (currentTime, departure, path) {
  const lastTime = Store.lastTime
  Store.lastTime = currentTime
  const currentDate = getCurrentDate()
  const updateTarget =  ['records', currentDate, path, 'acTime']
  if (!_.has(Store, updateTarget)) {
    _.set(Store, updateTarget, [0])
  }
  const interval = currentTime - lastTime
  const acTime = _.get(Store, updateTarget)
  if (interval > departure) {
    acTime.push(0)
  } else {
    acTime[acTime.length - 1] += interval
  }
}

exports.getAcTime = function () {
  const currentDate = getCurrentDate()
  const getTarget = ['records', currentDate]
  let result = 0
  if (_.has(Store, getTarget)) {
    const map = _.get(Store, getTarget)
    const dirs = Object.keys(map)
    result = dirs.reduce(function (sum, value) {
      const itemSum = map[value].acTime.reduce(function (s, item) {
        return s + item
      }, 0)
      return sum + itemSum
    }, result)
  }
  return result
}

exports.get = function (key) {
  return Store[key]
}

exports.set = function (key, value) {
  Store[key] = value
}


src/controller.js
const { window, StatusBarAlignment, Disposable, workspace } = require('vscode')
const { get, set, updateLastTime, getAcTime } = require('./store')
const { getCurrentTime, reverseTime } = require('./utils')
const locale = require('./locale')

const EXTENSIONNAME = "rest"

class RestController {
  constructor () {
    this.initConfig()
    this.bindEvent()
    this.startTime = getCurrentTime()
    this.remindTime = this.startTime + this._interval
    this._getAllRootPath(workspace.workspaceFolders)
    this.statusItem = this._createStatusItem(4)
  }

  // init config
  initConfig () {
    this._config = workspace.getConfiguration(EXTENSIONNAME)
    if (!this._config) return
    this._interval = reverseTime(this._config.interval, true)
    if (!this._interval) return
    this._locale = this._config.locale
    locale.init(this._locale)
    if (!this._locale) return
    this._departure = reverseTime(this._config.departure, true)
    if (!this._departure) return
    this._isShowTips = this._config.isShowTips
  }

  /**
   * bind all events which trigger to update state
   */
  bindEvent () {
    let subscriptions = []
    window.onDidChangeTextEditorSelection(this.onUpdateState, this, subscriptions)
    window.onDidChangeActiveTextEditor(this.onUpdateState, this, subscriptions)
    this._disposable = Disposable.from(...subscriptions)
  }

  _getAllRootPath (workspaceFolders = []) {
    let folders = []
    workspaceFolders.forEach(function (folder) {
      folders.push(folder.uri.fsPath)
    })
    this.workspaceFolders = folders
  }
  _getCurrentRootPath (path) {
    let resultPath = path
    for (let i = 0; i < this.workspaceFolders.length; i += 1) {
      let nowPath = this.workspaceFolders[i]
      if (path.indexOf(nowPath) >= 0) {
        resultPath = nowPath
        break;
      }
    }
    return resultPath
  }

  _getCurrentPath (event) {
    let documentData = event.textEditor
    ? event.textEditor._documentData
    : event._documentData
    if (documentData) {
      return this._getCurrentRootPath(documentData._uri.fsPath)
    }
    return ''
  }
  /**
   * update extension state
   */
  onUpdateState (event) {
    if (!event) return
    const currentRoot = this._getCurrentPath(event)
    if (!currentRoot) return
    const lastTime = get('lastTime')
    const currentTime = getCurrentTime()
    updateLastTime(currentTime, this._departure, currentRoot)
    const accTime = getAcTime()
    let conTime
    if (currentTime - lastTime > this._departure) {
      this.startTime = currentTime
      this.remindTime = currentTime + this._interval
      conTime = 0
    } else {
      conTime = currentTime - this.startTime
      // remind user have a break
      if (currentTime >= this.remindTime && this._isShowTips) {
        this.remindTime = currentTime + this._interval
        this._showTips('Warning', conTime, 'tip')
      }
    }
    this._updateStatusItem(conTime, accTime)
  }

  _getContiuousTime () {
    const currentTime = getCurrentTime()
    const lastTime = get('lastTime')
    if (currentTime - lastTime > this._departure) {
      return 0
    } else {
      return currentTime - this.startTime
    }
  }
  showContinuousTime () {
    this._showTips('Information', this._getContiuousTime(), 'statusItemCon')
  }

  showAccumulatedTime () {
    this._showTips('Information', getAcTime(), 'statusItemAcc')
  }
  /**
   * dispose all the subscriptions
   */
  dispose () {
    this._disposable.dispose()
  }

  // show user a information message
  _showTips (type, timestamp, item) {
    const msg = locale.get(item, this._transformTime(timestamp))
    window[`show${type}Message`](msg)
  }

  // create a status bar item
  _createStatusItem (priority) {
    return window.createStatusBarItem(StatusBarAlignment.Left, priority)
  }

  // transform timestamp to user-friendly format
  _transformTime (timestamp) {
    const munit = locale.get('minutes')
    const hunit = locale.get('hours')
    const minutes = Math.floor(timestamp / 60000)
    let h = ''
    let hn = Math.floor(minutes / 60)
    let m = ''
    let mn = minutes % 60
    h = '' + hn + hunit + ' '
    if (hn === 0) h = ''
    m = '' + mn + munit
    if (hn !== 0 && mn === 0) m = ''
    if (hn === 0 && mn === 0) m = locale.get('less') + ' 1' + munit
    return h + m
  }
  // update status item
  _updateStatusItem (ctime, atime) {
    const aText = locale.get('statusItemAcc', this._transformTime(atime))
    const cText = locale.get('statusItemCon', this._transformTime(ctime))
    this.statusItem.text = `${cText}  ${aText}`
    this.statusItem.show()
  }
}


exports.RestController = RestController
